module/apmfasthttp/server.go

```diff
@@ -93,6 +93,12 @@ func (h *apmHandler) handler(ctx *fasthttp.RequestCtx) {
 	}()
 
 	h.requestHandler(ctx)
+
+	if err := setResponseContext(ctx, tx, bc); err != nil {
+		return
+	}
+
+	tx.End()
```

Comment:
Isn't it a problem that in the error case the transaction is not ended?

Reply:
Also my understanding from the original PR is that there are cases (streaming) where the handler may return before the request is done. I was led to believe the closer would be invoked when the request completed. Is that not happening?

Reply:
As far as I could see in the original code, the closer was not being invoked. When I inlined `setResponseContext`, I started getting the correct status code in my test case. When I saw that inlining the code "worked" (without knowing about early returns from streaming cases), `txCloser` looked unnecessary to me since it was never being used. But I guess that's what the code in `init` was supposed to accomplish?

Reply:
Not sure what `init` you're referring to, but my understanding is that the closer added by StartTransactionWithBody is meant to be invoked when the request ends:

https://pkg.go.dev/github.com/valyala/fasthttp#RequestCtx.SetUserValue

> All the values are removed from ctx after returning from the top RequestHandler. Additionally, Close method is called on each value implementing io.Closer before removing the value from ctx.

I suppose that's not happening as expected

Reply:
The `init()` func: https://github.com/elastic/apm-agent-go/pull/1104/files#diff-68d60b1d06492503cf46894612cd31efc1d0e3ef6456a202af9a162e52f6e9b8L36-L54

>I suppose that's not happening as expected

According to my test case, it wasn't :shrug:

Reply:
I did a bit of testing and digging, and found that the closer is called but it's called after the response is reset - at which point the status code, headers, etc. are lost.

The reason for the closer is due to https://pkg.go.dev/github.com/valyala/fasthttp#RequestCtx.SetBodyStreamWriter. The handler may return but have set a streaming body writer; the request does not actually complete until after the body writer returns.

The docs for that method say:

> Access to RequestCtx and/or its' members is forbidden from sw.

So I think what we should do is:
 - set response context at the end of the handler wrapper
 - keep the closer, but use it only for ending the transaction and discarding the body capturer

I think we still need the `init` code. `fasthttp.RequestCtx` implements `context.Context` directly, with `Value` returning things set with `SetUserValue`: https://github.com/valyala/fasthttp/blob/v1.29.0/server.go#L2640